pftools/python/parflow/tools/io.py

```diff
@@ -21,76 +26,684 @@
+            x0 = 0 if not x0 else x0[0]
+            x1 = np.flatnonzero(end == coords)
+            x1 = None if x1 is None or len(x1) == 0 else x1[0]
+            return slice(x0, x1)
+
+        def _get_needed_subgrids(start, end, coords):
+            """ Helper function to clean up subgrid selection """
+            for s, c in enumerate(coords):
+                if start in c: break
+            for e, c in enumerate(coords):
+                if end in c: break
+            return np.arange(s, e+1)
+
+        if not nz:
+            nz = self.header['nz']
+        end_x = start_x + nx
+        end_y = start_y + ny
+        end_z = start_z + nz
+        p, q, r = self.header['p'], self.header['q'], self.header['r']
+        # Convert to numpy array for simpler indexing
+        x_coords = np.array(self.coords['x'], dtype=object)
+        y_coords = np.array(self.coords['y'], dtype=object)
+        z_coords = np.array(self.coords['z'], dtype=object)
+
+        # Determine which subgrids we need to read
+        p_subgrids = _get_needed_subgrids(start_x, end_x, x_coords)
+        q_subgrids = _get_needed_subgrids(start_y, end_y, y_coords)
+        r_subgrids = _get_needed_subgrids(start_z, end_z, z_coords)
+
+        # Determine the coordinates of these subgrids
+        x_sg_coords = np.unique(np.hstack(x_coords[p_subgrids]))
+        y_sg_coords = np.unique(np.hstack(y_coords[q_subgrids]))
+        z_sg_coords = np.unique(np.hstack(z_coords[r_subgrids]))
+        # Min values will be used to align in the bounding data
+        x_min = np.min(x_sg_coords)
+        y_min = np.min(y_sg_coords)
+        z_min = np.min(z_sg_coords)
+        # Make an array which can fit all of the subgrids
+        full_size = (len(x_sg_coords), len(y_sg_coords), len(z_sg_coords))
+        bounding_data = np.empty(full_size, dtype=np.float64)
+        subgrid_iter = itertools.product(p_subgrids, q_subgrids, r_subgrids)
+        for (xsg, ysg, zsg) in subgrid_iter:
+            subgrid_idx = xsg + (p * ysg) + (p * q * zsg)
+            # Set up the indices to insert subgrid data into the bounding data
+            x0, y0, z0 = self.subgrid_start_indices[subgrid_idx]
+            x0, y0, z0 = x0 - x_min, y0 - y_min, z0 - z_min
+            dx, dy, dz = self.subgrid_shapes[subgrid_idx]
+            x1, y1, z1 = x0 + dx, y0 + dy, z0+ dz
+            bounding_data[x0:x1, y0:y1, z0:z1] = self.iloc_subgrid(subgrid_idx)
+
+        # Now clip out the exact part from the bounding box
+        clip_x = _get_final_clip(start_x, end_x, x_sg_coords)
+        clip_y = _get_final_clip(start_y, end_y, y_sg_coords)
+        clip_z = _get_final_clip(start_z, end_z, z_sg_coords)
+        if z_first:
+            ret_data = bounding_data[clip_x, clip_y, clip_z].T
+        else:
+            ret_data = bounding_data[clip_x, clip_y, clip_z]
+        return ret_data
+
+
+    def loc_subgrid(self, pp: int, qq: int, rr: int) -> np.ndarray:
```

Comment:
I agree it's a bit of a tricky thing to maintain naming conventions across languages. To maintain consistency/traceability in the discussion I'm adding my email response here:

> 
> My position on this is (selfishly) that I would really rather not spend a lot of time on refactoring the reader code to use the C conventions. Normally a refactor would be no problem, but these are numerous variables with tiny names and case sensitivity, which is just begging for bugs to be introduced IMHO - not to mention it would take a decent amount of time that I would guess none of us really want to give up. In light of that, I've got a proposed alternative. Right now the main discrepancies between the Parflow C code and our python code is:
>  
> ```
> Description              | Parflow C        | Current Python  | Proposed Python        
> -------------------------+------------------+-----------------+-------------------
> Subgrid lengths          | P, Q, R          | p, q, r         | p, q, r
> Subgrid indices          | p, q, r          | pp, qq, rr      | sg_p, sg_q, sg_r
> Size of the full grid    | NX, NY, NZ       | nx, ny, nz      | nx, ny, nz
> Size of a subgrid        | nx, ny, nz       | snx, sny, snz   | sg_nx, sg_ny, sg_nz
> Cell size of the grid    | dx, dy, dz       | dx, dy, dz      | dx, dy, dz           (All the same)
> Index into subgrid cell  | i, j, k          | Doesn't exist   | N/A                  (No analogous concept)
> Lower left index         | ix, iy, iz       | ix, iy, iz      | ix, iy, iz           (All the same)
> Stride in index space    | sx, sy, sz       | Doesn't exist   | N/A                  (No analogous concept)
> Index into PF vectors    | ip               | Doesn't exist   | N/A                  (No analogous concept)
> ```
> 
> In summary, I'm proposing that any subgrid-specific quantities in the python code use `sg_` as a prefix, and then the analogous name to the fullgrid name if it exists. This proposal also has the added benefit that all variable names are kept lower-case, maintaining common python code conventions.